bibtex-actions.el

```diff
@@ -229,26 +280,36 @@ offering the selection candidates"
              collect (cdr (or (assoc choice candidates)
                               (rassoc choice candidates))))))
 
-(defun bibtex-actions--normalize-paths (file-paths)
-  "Return a list of FILE-PATHS normalized with truename."
-  (if (stringp file-paths)
-      ;; If path is a string, return as a list.
-      (list (file-truename file-paths))
-    (delete-dups (mapcar (lambda (p) (file-truename p)) file-paths))))
-
 (defun bibtex-actions--local-files-to-cache ()
   "The local bibliographic files not included in the global bibliography."
   ;; We cache these locally to the buffer.
   (let* ((local-bib-files
           (bibtex-actions--normalize-paths
-           (bibtex-completion-find-local-bibliography)))
-         (global-bib-files
-          (bibtex-actions--normalize-paths
-           bibtex-completion-bibliography)))
-    (seq-difference local-bib-files global-bib-files)))
-
-(defun bibtex-actions--format-candidates (&optional context)
-  "Format candidates, with optional hidden CONTEXT metadata.
+           (bibtex-completion-find-local-bibliography))))
+    (seq-difference local-bib-files bibtex-actions-bibliography)))
+
+(defun bibtex-actions-get-value (field item &optional _default)
+  "Return biblatex FIELD value for ITEM."
+  (or (cdr (assoc-string field item 'case-fold))
+      (cl-loop for fname in (cdr (assoc field bibtex-actions-field-map))
+               when (cdr (assoc-string fname item 'case-fold))
+                         return (cdr (assoc-string fname item 'case-fold)))
+      ""))
```

Comment:
Why are we returning empty string instead of `nil` of the value is not found? It seems strange and as it is is causing the candidates to display as if all of them have a pdf.

And the optional `_default` argument is unused. Maybe both are related?

Reply:
> Why are we returning empty string instead of nil of the value is not found?

I can try again, but I think I ran into errors when I didn't do this.

> It seems strange and as it is is causing the candidates to display as if all of them have related files.

That's weird; I don't see that.

![image](https://user-images.githubusercontent.com/1134/129491118-41f3724c-f005-4a47-80f5-a6ae1c93171e.png)

> And the optional _default argument is unused.

I think that's a vestige of what started out as a BC function.

Reply:
> I can try again, but I think I ran into errors when I didn't do this.

I did, but it seems a bug in `bibtex-actions-shorten-names`:

```console
bibtex-actions-shorten-names: Wrong type argument: characterp, ""
```

Reply:
> That's weird; I don't see that.

Maybe evaluate `bibtex-actions--format-candidates` again,  refresh the cache and then check. It wouldn't show until a cache refresh

Reply:
What's really weird is now I see this, after re-adding the `""`.

> It seems strange and as it is is causing the candidates to display as if all of them have related files.

I am guessing the problem is in the format-entry function.

Reply:
> I am guessing the problem is in the format-entry function.

Yes, it seems like there a string is always expected. One way of dealing with that would be to replace `""` with "_default" and then pass the empty string in the format-entry function

Reply:
> Yes, it seems like there a string is always expected. One way of dealing with that would be to replace `""` with `_default` and then pass the empty string in the format-entry function.

I'm not seeing where to do the latter ATM, as I'm looking at BC.

Reply:
> I'll look at it again tomorrow. It's probably something obvious.

Sounds good. 

I am not seeing any errors but I have an example of a bib entry causing them, send them my way, though I will mostly disappear from tomorrow.

Reply:
I may have just fixed it, and it was indeed a simple thing.